### PR TITLE
TOF powder diffraction GUI 

### PR DESCRIPTION
**Pull request by koshchii**
**Description of work.**

This PR implements a GUI for the TOF powder diffraction data reduction workflow at DNS.

**To test:**

- Download test data: [TOF_Powder_Vanadium.zip](https://github.com/mantidproject/mantid/files/8629427/TOF_Powder_Vanadium.zip)
- Open the reduction GUI [Interfaces->Direct->DNS Reduction]
- Browse to the directory where the test data are located using the "Paths" tab of the GUI
- Load and select the data for reduction using the "Data" tab
- Choose the data reduction options using the "Options" tab
- Reduce the data and generate a data reduction script using the "Script Generator" tab
- Visualize results using the "Plotting" tab 

Resolves Part 1a) of #27538. 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by Harrietbrown** [comment]
on opening DNS Reduction I get this error

```
Traceback (most recent call last):
  File "c:\users\wey38795\documents\mantid_builds\master\raw\mantid\qt\applications\workbench\workbench\app\mainwindow.py", line 402, in <lambda>
    lambda checked_py, script=script: self.launch_custom_python_gui(script))
  File "c:\users\wey38795\documents\mantid_builds\master\raw\mantid\qt\applications\workbench\workbench\app\mainwindow.py", line 360, in launch_custom_python_gui
    self.interface_executor.execute(open(filename).read(), filename)
  File "c:\users\wey38795\documents\mantid_builds\master\raw\mantid\qt\python\mantidqt\mantidqt\widgets\codeeditor\execution.py", line 150, in execute
    executor.execute(code_str, filename, flags, self.globals_ns, line_offset)
  File "c:\users\wey38795\documents\mantid_builds\master\raw\mantid\qt\python\mantidqtinterfaces\mantidqtinterfaces\DNS_Reduction.py", line 38
    app.setWindowIcon(QtGui.QIcon(f'{app_dir}/dns_powder_tof/dns_icon.png'))
                                  ^
SyntaxError: Format strings are only supported in Python 3.6 and greater
```

**Comment by thomueller** [comment]
I am on 3.8.13, and there is no error. If you tell me which version of 3.8 you are using I could try it out.

**Comment by Harrietbrown** [comment]
I'm running 3.8.0

**Comment by thomueller** [comment]
This has to do with the way mantid executes the python code of the custom interface.
There was a bug in python which was fixed in 8.3.4:
https://bugs.python.org/issue35975#msg371938

**Comment by martyngigg** [comment]
> I think this has to do with the way mantid executes the python code of the custom interface. There was a bug in python which was fixed in 3.8.4: https://bugs.python.org/issue35975#msg371938

I just got pinged about this conversation. Thanks for the heads up. It looks like we should solve this once we use our Conda-based builds for the release and building as we will get the latest version of Python 3.8.

**Comment by martyngigg** [comment]
The branch for release 6.4 has been created so my default I've moved this to 6.5 so it would now first end up in the nightly. As this is quite a large set of changes there is some time needed for review and we need to make sure that everything going on to 6.4 is starting to be smaller bug fixes now.

**Comment by thomueller** [comment]
Hi Martyn, sure we did not expect that it could be reviewed in a day. Submitting it before the code freeze was more a deadline to us.  And we should test it very well.
We tried to make the initial submit as small as possible, by submitting only one reduction mode of the GUI, but it is still quite large.

**Comment by mganeva** [comment]
I suggest Milestone 6.5 S1 as a compromise. However, if the PR is too large to review it properly by the 1st of July, feel free to shift the Milestone to the next one.

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/data_structures/dns_file.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+    This is a dictionary, but can also be accessed like attributes.
+    """
+    def __init__(self, datapath, filename):
+        super().__init__()
+        self.new_format = self.read(datapath, filename)
+
+    def write(self, datapath, filename):
+        # mostly stolen form nicos
+        txt = ''
+        separator = "#" + "-" * 74 + "\n"
```
Comment (Harrietbrown) [inline qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/data_structures/dns_file.py:29]: Is the DNS file formate something we have produced ourselves? much of this looks like it would be better handled as a JSON file instead of a long txt file. and would be less vulnurable to user error.

File: docs/source/interfaces/direct/dns_powder_tof/DNS Powder TOF.rst [hunk 2]
```diff
@@ -0,0 +1,38 @@
+.. _dns_powder_tof-ref:
+
+.. toctree::
```
Comment (Harrietbrown) [inline docs/source/interfaces/direct/dns_powder_tof/DNS Powder TOF.rst:3]: The help option in the interface is not pointing to this doc. I think it needs to be moved.

File: docs/source/interfaces/direct/dns_powder_tof/dns_powder_tof_options_tab.rst [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+binning and proceed to the next tab if no change to the selected options is
+desired.
+
+The **Wavelength** value can be overwritten, for example for the cases when
+:math:`\lambda/2` was selected by the velocity selector.
+
+**Detector Efficiency Correction** norms the detector efficiency using standard
+vanadium data.
+
+Instrument background can be subtracted from vanadium and sample data,
```
Comment (Harrietbrown) [inline docs/source/interfaces/direct/dns_powder_tof/dns_powder_tof_options_tab.rst:22]: Each entry here should begin with the name of the option in the gui, e.g. `**Subtract Instrument Background from Vanadium** can be enabled to subtract the instrument background from vanadium and sample data.`

Also does the subtraction happen automatically from the sample data or does it have to be enabled through this option, if so the name of the option should be changed to show that it also subtracts from the sample.

File: docs/source/interfaces/direct/dns_powder_tof/dns_powder_tof_script_generator_tab.rst [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+   :align: center
+   :height: 400px
+\
+
+This tab enables a user to generate and run a mantid script for data reduction
+by clicking on the **Generate Script** button.
+
+**Automatic Script Filename** checkbox is selected by default and the output
+script filename in this case will be set automatically to
+``script_samplename_from_firstfilenumber_to_lastfilenumber.py``. Alternatively,
```
Comment (Harrietbrown) [inline docs/source/interfaces/direct/dns_powder_tof/dns_powder_tof_script_generator_tab.rst:16]:
```suggestion
``script_{sample name}_from_{first file number}_to_{last file number}.py``. Alternatively,
```

File: docs/source/interfaces/direct/dns_powder_tof/dns_powder_tof_plotting_tab.rst [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+
+DNS Powder TOF Plotting Tab
+---------------------------
+
+.. image::  ../../../images/DNS_interface_powder_tof_plotting_tab.png
+   :align: center
+   :height: 400px
+\
+
+The plotting tab of the powder TOF mode, is just the Mantid :ref:`sliceviewer`.
```
Comment (Harrietbrown) [inline docs/source/interfaces/direct/dns_powder_tof/dns_powder_tof_plotting_tab.rst:11]:
```suggestion
For usage of the Powder TOF plotting tab see usage for the mantid :ref:`sliceviewer`.
```

**Review by Harrietbrown — CHANGES_REQUESTED** [review]
The line indents on the generated script are not consistent. it would be good if it could be made to have that. e.g.
From this:
image
To this:
image



Bug found:

1. Set the Data Directory
2. Select all sample runs, vanadium and empty
3. in options check `Mask Bad Detectors`
4. click script generator.
5. You get an error caused by numpy not being imported.

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/data_structures/dns_tof_powder_dataset.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+        dic = self.data_dic.get(sample_type, 0)
+        if dic:
+            return len(dic.keys()) - 1
+        return 0
+
+    def get_vana_filename(self):
+        vana_filename = [x for x in self.data_dic.keys() if '_vana' in x]
+        if len(vana_filename) == 0:
+            vana_filename = ''
+        elif len(vana_filename) > 1:
```
Comment (Harrietbrown) [inline qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/data_structures/dns_tof_powder_dataset.py:137]: Am I right to assume that there should never be more than one Vanadium filename? if so should there be a error message if two are found?

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/data_structures/dns_treemodel.py [hunk 7]
```diff
@@ -0,0 +1,394 @@
...
+        return (self.scan_from_row(row).childCount()
+                >= self._scan_expected_points_from_row(row))
+
+    def text_in_scan(self, row, text):
+        return text in self._scan_command_from_row(row)
+
+    def _get_last_row(self):
+        return self.number_of_scans() - 1
+
+    def data(self, index, role):  # overrides QT function
```
Comment (Harrietbrown) [inline qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/data_structures/dns_treemodel.py:119]: could the index be stored as a enum so it is clear what is being called for?

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/data_structures/dns_treemodel.py [hunk 8]
```diff
@@ -0,0 +1,394 @@
...
+        checked = self.match(self.index(0, 0, QModelIndex()),
+                             Qt.CheckStateRole, Qt.Checked, -1,
+                             Qt.MatchExactly | Qt.MatchRecursive)
+        n_checked = []
+        for index in checked:
+            item = self._item_from_index(index)
+            if not item.hasChildren():
+                if full_info:
+                    n_checked.append({
+                        'file_number': int(item.data(0)),
```
Comment (Harrietbrown) [inline qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/data_structures/dns_treemodel.py:199]: storing the contents of each column as a enum will make these calls clearer.

**Review by Harrietbrown — CHANGES_REQUESTED** [review]

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/file_selector/file_selector_presenter.py [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+                 model=None,
+                 watcher=None):
+        # pylint: disable=too-many-arguments
+        super().__init__(parent=parent, name=name, view=view, model=model)
+        self.watcher = watcher
+        self.view = view
+        # set sample data view
+        self.view.set_tree_model(self.model.get_model())
+        # set standard data view
+        self.view.set_tree_model(self.model.get_model(standard=True),
```
Comment (Harrietbrown) [inline qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/file_selector/file_selector_presenter.py:30]: It would be better to have view have two functions, one to set the sample tree and one to set the standard tree

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/data_structures/dns_file.py (hunk 1 above)
Comment (koshchii) [reply qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/data_structures/dns_file.py:29]: That's a good point - thanks! Briefly speaking, we don't produce the DNS file format. This format is generated by an "external" Nicos software and it has been used for many years at DNS as a format to store raw experimental data. The GUI takes (loads) ".d_dat" files as an input and then a user chooses the options for data reduction and then the corresponding algorithms are executed.   

To answer your question, the write() method here was included to be used for another (Powder Elastic) operational mode of the GUI that has not been submitted yet. In that mode, we had an option to perform interpolation of standard files for the case when standard and sample data are measured at different angles. And the interpolation algorithm would use the write() method to create "interpolated standard datafiles". However, recently we decided that generating "interpolated datafiles" is a dangerous approach and we won't use it in the Powder Elastic mode. So I will remove this method upon revision.

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/file_selector/file_selector_presenter.py [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+                                 standard=True)
+
+        # adjust view to column width
+        self.num_columns = self.model.get_active_model_column_count()
+        self.view.adjust_treeview_columns_width(self.num_columns)
+
+        self._old_data_set = set()
+
+        # connect signals
+        self.view.sig_expanded.connect(self._expanded)
```
Comment (Harrietbrown) [inline qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/file_selector/file_selector_presenter.py:40]: Could you move the signal conectors into their own function?

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/file_selector/file_selector_presenter.py [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+        self.view.sig_dataset_changed.connect(self._dataset_changed)
+        self.watcher.sig_files_changed.connect(self._files_changed_by_watcher)
+
+    # loading
+    def _read_all(self, filtered=False, watcher=False, start=None, end=None):
+        """
+        Reading of new files, if filtered is True, only the files in the
+        range specified by start and end.
+        """
+        fn_range = [start, end]
```
Comment (Harrietbrown) [inline qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/file_selector/file_selector_presenter.py:62]: What does fn stand for?

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/file_selector/file_selector_presenter.py [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+        data_path = self.param_dict['paths']['data_dir']
+        nb_files, loaded, datafiles, fn_range = \
```
Comment (Harrietbrown) [inline qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/file_selector/file_selector_presenter.py:64]: What does nb stand for?

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/file_selector/file_selector_presenter.py [hunk 13]
```diff
@@ -0,0 +1,281 @@
...
+            self.watcher.stop_watcher()
+
+    def _auto_select_standard(self, state):
+        if state == 2:
+            self._check_all_visible_scans()
+
+    # scan selection
+    def _automatic_select_all_standard_files(self):
+        was = not self.model.model_is_standard()
+        if was:  # if view is not standard we change to it and change back
```
Comment (Harrietbrown) [inline qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/file_selector/file_selector_presenter.py:130]: Why is this required

**Review by Harrietbrown — CHANGES_REQUESTED** [review]

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/script_generator/tof_powder_script_generator_model.py [hunk 14]
```diff
@@ -0,0 +1,333 @@
...
+DNS script generator for TOF powder data
+"""
+
+from mantidqtinterfaces.dns_powder_tof.data_structures.dns_tof_powder_dataset \
+    import DNSTofDataset
+from mantidqtinterfaces.dns_powder_tof.script_generator.common_script_generator_model \
+    import DNSScriptGeneratorModel
+
+
+class DNSTofPowderScriptGeneratorModel(DNSScriptGeneratorModel):
```
Comment (Harrietbrown) [inline qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/script_generator/tof_powder_script_generator_model.py:17]: I can see how much you have done on this p[rocess, but how much of it could be handled running the functions and using the workspace fistory?

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/data_structures/dns_tof_powder_dataset.py [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+                inside = True
+                break
+            inside = False
+        if not inside:
+            new_arr.append(bank)
+    return new_arr
+
+
+def _get_max_key_length(dataset):
+    return max([len(a) for a in dataset.keys()]) + 6 + 4
```
[the remark this answers is not in the document]
Comment (koshchii) [reply qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/data_structures/dns_tof_powder_dataset.py:40]: This method was implemented for alignment purposes inside the generated script, so 4 extra spaces were needed to account for a tab and 6 extra spaces were needed to account for special characters like ', :, ect. Since you requested a modification of the indentation style, this method won't be needed any longer and will be removed on resubmission.

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/xml_dump/xml_dump_widget.py [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+    DNSWidget
+from mantidqtinterfaces.dns_powder_tof.xml_dump.xml_dump_model import \
+    DNSXMLDumpModel
+from mantidqtinterfaces.dns_powder_tof.xml_dump.xml_dump_presenter import \
+    DNSXMLDumpPresenter
+from mantidqtinterfaces.dns_powder_tof.xml_dump.xml_dump_view import \
+    DNSXMLDumpView
+
+
+class DNSXMLDumpWidget(DNSWidget):
```
Comment (Harrietbrown) [inline qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/xml_dump/xml_dump_widget.py:22]: I am not seeing this widget in the GUI, it is enabled currently?

**Review by Harrietbrown — CHANGES_REQUESTED** [review]

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/data_structures/dns_tof_powder_dataset.py (hunk 6 above)
Comment (koshchii) [reply qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/data_structures/dns_tof_powder_dataset.py:137]: Good point, thanks! The error message will be modified to mention that only one vanadium "filename" can be selected if the vanadium correction is chosen to be calculated. Actually, the naming of this variable (and the method where this variable is defined) is misleading, because it describes rather the name of the scan than the filename. Several vanadium files can be chosen, but they have to belong to the same scan. So this method and the variable will be refactored

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/file_selector/file_selector_model.py [hunk 17]
```diff
@@ -0,0 +1,278 @@
...
+from mantidqtinterfaces.dns_powder_tof.helpers.file_processing import (
+    filter_filenames, load_txt, open_editor, return_filelist, save_txt,
+    unzip_latest_standard)
+
+
+class DNSFileSelectorModel(DNSObsModel):
+    def __init__(self, parent=None):
+        super().__init__(parent)
+        # Model has this function to call back on parent.
+        # No idea how to avoid this.
```
[the remark this answers is not in the document]
Comment (koshchii) [reply qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/file_selector/file_selector_model.py:25]: Right, the file selector model should be informed by the presenter about the progress and widget does it for the presenter. I'll remove this comment on re-submission

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/file_selector/file_selector_model.py [hunk 18]
```diff
@@ -0,0 +1,278 @@
...
+        self.update_progress = parent.update_progress
+        # Qt tree models
+        self.tree_model = DNSTreeModel()
```
[the remark this answers is not in the document]
Comment (koshchii) [reply qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/file_selector/file_selector_model.py:28]: Right, I will refactor the names of these models so that they are more descriptive

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/file_selector/file_selector_presenter.py (hunk 9 above)
Comment (koshchii) [reply qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/file_selector/file_selector_presenter.py:30]: Right, thanks for the suggestion! Will fix that

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/file_selector/file_selector_presenter.py (hunk 11 above)
Comment (koshchii) [reply qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/file_selector/file_selector_presenter.py:62]: It stands for "file number", I'll refactor it

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/file_selector/file_selector_presenter.py (hunk 12 above)
Comment (koshchii) [reply qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/file_selector/file_selector_presenter.py:64]: It stands for "number of files". It will be refactored, as well.

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/file_selector/file_selector_presenter.py (hunk 13 above)
Comment (koshchii) [reply qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/file_selector/file_selector_presenter.py:130]: Thanks, it is not necessary since process_request() calls this method it in the "sample mode" anyways.

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/main_view.py [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+        # connect menu signals
+        self.ui.actionQuit.triggered.connect(self.close)
+        self.ui.actionSave_as.triggered.connect(self._save_as_triggered)
+        self.ui.actionSave.triggered.connect(self._save_triggered)
+        self.ui.actionOpen.triggered.connect(self._open_triggered)
+        self.ui.actionMantid_help.triggered.connect(self._help_button_clicked)
+        self.ui.actionDNS_website.triggered.connect(self._open_dns_webpage)
+        # connect mode switching signals
+        self.modus_mapping = {
+            # self.ui.actionSimulation: 'simulation',
```
[the remark this answers is not in the document]
Comment (koshchii) [reply qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/main_view.py:47]: They will be used when the other modes of the GUI are implemented, but they can be removed in this PR

File: docs/source/interfaces/direct/dns_powder_tof/dns_powder_tof_options_tab.rst (hunk 3 above)
Comment (koshchii) [reply docs/source/interfaces/direct/dns_powder_tof/dns_powder_tof_options_tab.rst:22]: > Also does the subtraction happen automatically from the sample data or does it have to be enabled through this option, if so the name of the option should be changed to show that it also subtracts from the sample.

The documentation for this option is misleading. This check mark can be used to enable background subtraction from vanadium data only. The documentation will be updated correspondingly

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/main_view.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (Harrietbrown) [reply qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/main_view.py:47]: It's ok to leave them in, just with a comment mentioning it.

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/data_structures/dns_obs_model.py [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+#     & Institut Laue - Langevin
+# SPDX - License - Identifier: GPL - 3.0 +
+
+"""
+DNS Observer Class, they share a common data model DNSReductionGui_model
+and are updated by DNSReductionGui_presenter.
+"""
+
+
+class DNSObsModel:
```
[the remark this answers is not in the document]
Comment (koshchii) [reply qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/data_structures/dns_obs_model.py:13]: I decided to keep it because this model is used by common_script_generator_model, which uses parent and contains non-static models

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/data_structures/dns_observer.py [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+        self.own_dict = OrderedDict()
+        self.modus = ''
+        self.request_from_abo = None
+
+    def update(self, param_dict):
+        """
+        Updating the own dictionary from ParameterAbo.
+        """
+        if self.modus != self.parent.parent.modus.name:
+            self.modus = self.parent.parent.modus.name
```
[the remark this answers is not in the document]
Comment (koshchii) [reply qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/data_structures/dns_observer.py:53]: DNSObserver's parents are various widgets (like DNSPathWidget, DNSFileSelectorWidget, etc), whereas the parent's parent is always DNSReductionGuiWidget where the modus is set, so the modus name is taken from there

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/data_structures/dns_observer.py [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+    def __init__(self, parent=None, name=None, view=None, model=None):
+        super().__init__()
+        self.name = name
+        self.parent = parent
+        self.view = view
+        self.model = model
+        self.param_dict = OrderedDict()
+        self.own_dict = OrderedDict()
+        self.modus = ''
+        self.request_from_abo = None
```
[the remark this answers is not in the document]
Comment (koshchii) [reply qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/data_structures/dns_observer.py:46]: ParameterAbo class serves to handle communication between different widgets through an observer pattern. Presenters are the observers and they are instances of DNSObserver class. They have an argument name which is a unique string and a method update which can get information from the GUI parameters. For automatic reduction, observers can request data processing from parameter abo through observer.request_from_abo = self.process_request

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/data_structures/dns_treemodel.py (hunk 7 above)
Comment (koshchii) [reply qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/data_structures/dns_treemodel.py:123]: I left it unchanged because I don’t see a straightforward way to implement that. Here an abstract tree model for accessing and manipulating data is implemented. In this model, we use a pointer-based tree structure and when we pass a model index to a view, we record the address of the corresponding item in the index and retrieve it later with internalPointer(). One model index may refer to several items, so I am not sure how to implement an enum here.

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/script_generator/tof_powder_script_generator_model.py (hunk 14 above)
Comment (koshchii) [reply qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/script_generator/tof_powder_script_generator_model.py:18]: I see your point. I am not sure why it was decided to not use the workspace history for the script generation purposes since I joined at the final stages of the GUI's development

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/xml_dump/xml_dump_widget.py (hunk 16 above)
Comment (koshchii) [reply qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/xml_dump/xml_dump_widget.py:22]: All widgets (children of the DNSWidget class) are considered as “containers” for presenter, model and view. The view of xml_dump is inherited from the DNSReductionGUIView class that loads dns_gui_main_reduced_menu.ui. The xml_dump then handles saving the GUI’s status to an xml document, as well as to subsequent loading of the file in order to save some time for a user.

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/data_structures/dns_observer.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (Harrietbrown) [reply qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/data_structures/dns_observer.py:53]: If all these widgets have the same base class you could have a function within that to return the parents modus.

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/script_generator/tof_powder_script_generator_model.py (hunk 14 above)
Comment (Harrietbrown) [reply qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/script_generator/tof_powder_script_generator_model.py:18]: Do you think it would be possible to change it to utalize workspace history before the code freeze?

**Review by Harrietbrown — CHANGES_REQUESTED** [review]

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/script_generator/tof_powder_script_generator_model.py (hunk 14 above)
Comment (koshchii) [reply qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/script_generator/tof_powder_script_generator_model.py:18]: Would it be OK if we leave this to the subsequent PR which will contain another operational mode of the GUI? That mode also contains the script generator model, so it will need to be updated anyways. I am attending a workshop/school this week and the schedule is dense, so I don't have an opportunity to implement it this week

Comment (Harrietbrown) [reply qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/script_generator/tof_powder_script_generator_model.py:18]: That will be acceptable.

**Review by Harrietbrown — APPROVED** [review]
This looks good, the interface works as described. the only issue is the method of generating the script not utalising the workspace history of the output. but that will be addressed in a seperate PR. as such this is ready to be merged for release.
Good Job
